[FLINK-13982][runtime] Implement memory calculation logics

xintongsong opened the pull request:
## What is the purpose of the change
This PR is part of FLIP-49. It implements the new task executor memory configuration and calculation logics.

## Brief change log
- adc4fb2acd7ae0ab2a03cc1e847f0add38d63f48: Rename constant names for managed memory size and fraction config options. This is to avoid naming conflicts in the next commit.
- 97057237606f35581b0ac5b53c4e5807e0ea0376: Introduce task executor memory config options for FLIP-49.
- 119fa285da71d191dcd4ebbd34f8f668c446f8b6: Introduce arithmetic operations (add, subtract and multiply) for MemorySize. These are used for memory calculations in following commits.
- 3acadf1377e10c2e817702e32960ca71d9c62d5c: Introduce 'TaskExecutorResourceSpec' as the data structure for storing calculated memory sizes.
- 652135947866e0d8d10d5081f106213f30287898: Introduce 'TaskExecutorResourceUtils' for calculating memory sizes from configurations, as well as 'TaskExecutorResourceUtilsTest' that validates the memory calculation logics.
- 1ae6e405a644fb981e1066163688bc147338e583: Implement generating JVM parameters and dynamic configurations ('-D' options) from 'TaskExecutorResourceSpec' in 'TaskExecutorResourceUtils'. JVM parameters will be used for launching task executor JVM with calculated memory sizes. Dynamic configurations will be used for passing calculated memory sizes into task executors. 

## Verifying this change
This change added tests and can be verified as follows:
- Add 'TaskExecutorResourceUtilsTest' that validates:
  - Memory calculation logics
    - Success and failure cases
    - Backwards compatibility
  - Generating dynamic configurations
  - Generating JVM parameters

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

KarmaGYZ commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -0,0 +1,323 @@
+package org.apache.flink.runtime.clusterframework;
```
Move this line after the license, though its not a checkstyle issue.

xintongsong commented on flink-core/src/main/java/org/apache/flink/configuration/MemorySize.java:
```diff
@@ -115,6 +115,27 @@ public String toString() {
+
+	public MemorySize multiply(double multiplier) {
+		checkArgument(multiplier >= 0, "multiplier must be >= 0");
+		double product = this.bytes * multiplier;
```
MemorySize is essentially a size in bytes stored in long type. Converting from double type to long type should not be a problem. We might loose at most 1 byte in each multiply operation due to the rounding down error, which should be acceptable..

KarmaGYZ replied:
Agreed. 
Besides, we could change the type of `product` to long, to avoid potential precision issue in next line. There is no use case to treat that value as double.

xintongsong replied:
Changing `product` to `long` could cause overflow.

xintongsong replied:
True, the error can be up to 9 bits and thus should not be ignored. I think it's a good idea to use BigDecimal for safety.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,333 @@
+			.add(shuffleMemorySize)
+			.add(managedMemorySize);
+
+		final Tuple2<MemorySize, MemorySize> totalFlinkExternalMemorySizeTuple2 = deriveTotalFlinkExternalMemorySizes(config, totalFlinkMemorySize);
```
We may derive JVM metaspace and overhead either from total flink memory or total process memory, thus `deriveMetaSpaceAndOverHeadMemorySizes` may not be a good name for me.

I guess the confusion comes from understanding what is 'total flink external'? How about add some explanations and ascii arts in javadoc of this class, to explain the flip49 memory model?

WDYT?

KarmaGYZ replied:
Add javadoc LGTM. It will also enhance the maintainability of this mechanism.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,333 @@
+			return MemorySize.parse(config.getString(TaskManagerOptions.TOTAL_FLINK_MEMORY));
+		} else {
+			final long legacyHeapMemoryMB = config.getInteger(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY_MB);
+			return new MemorySize(legacyHeapMemoryMB << 20); // megabytes to bytes
```
`legacyHeapMemoryMB ` is initialized with an integer value and can be no larger than Integer.MAX_VALUE. Thus it cannot cause a long type overflow with '<< 20'.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,333 @@
+	private static RangeFraction getManagedMemoryRangeFraction(final Configuration config) {
+		final MemorySize minSize = new MemorySize(0);
+		final MemorySize maxSize = new MemorySize(Long.MAX_VALUE);
+		final double fraction = config.getFloat(TaskManagerOptions.MANAGED_MEMORY_FRACTION);
```
What happens if the user has `LEGACY_MANAGED_MEMORY_FRACTION` configured? Don't we want to support backwards compatibility?

xintongsong replied:
True. I forgot to add `LEGACY_MANAGED_MEMORY_FRACTION.key()` as a deprecated key to the new `MANAGED_MEMORY_FRACTION`

tillrohrmann replied:
I see. Could we transform `legacyFraction` into `flip49Fraction` if `LEGACY_MANAGED_MEMORY_FRACTION` has been specified and `MANAGED_MEMORY_FRACTION` has not been specified?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,333 @@
+	//  Memory Configuration Calculations
+	// ------------------------------------------------------------------------
+
+	public static TaskExecutorResourceSpec resourceSpecFromConfig(final Configuration config) throws ConfigurationException {
```
I think the way we calculate the resource requirements is quite hard to understand for a user. Depending on which values the user configured, some have precedence over others and some values will silently be ignored. For example, if task heap memory and managed memory is configured then Flink does not respect the total flink memory size. This is confusing in my opinion.

Instead I would suggest to do the following: We take all configured values and compute the missing values based on the input. The user's input should always have precedence otherwise it becomes hard to understand. Before returning the result, we do a sanity check that things actually add up. If not, then we fail and notify the user about the misconfigured setup. For example, if the users configures total process memory < total flink memory, then I would expect that we fail.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -348,6 +348,18 @@
+	public static final ConfigOption<String> MANAGED_MEMORY_OFFHEAP_SIZE =
+		key("taskmanager.memory.managed.off-heap.size")
+			.noDefaultValue()
+			.withDescription("DO NOT USE THIS CONFIG OPTION! This is the config option for Off-Heap Managed Memory size."
```
Why shouldn't it be used? Can't we do a sanity check and fail if things don't add up?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -37,6 +37,24 @@
+		final MemorySize jvmHeapSize = taskExecutorResourceSpec.getFrameworkHeapSize()
+			.add(taskExecutorResourceSpec.getTaskHeapSize())
+			.add(taskExecutorResourceSpec.getOnHeapManagedMemorySize());
+		final MemorySize jvmDirectSize = taskExecutorResourceSpec.getTaskOffHeapSize()
```
What about `getOffHeapManagedMemorySize`?

xintongsong replied:
That uses unsafe memory, which should not be accounted for either `-Xmx` nor` -XX:MaxDirectMemorySize`.

tillrohrmann replied:
Good point. Thanks for the clarification.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -0,0 +1,323 @@
+
+		assertThat(heapSizeMax, is(TM_RESOURCE_SPEC.getFrameworkHeapSize().add(TM_RESOURCE_SPEC.getTaskHeapSize()).add(TM_RESOURCE_SPEC.getOnHeapManagedMemorySize())));
+		assertThat(heapSizeMin, is(heapSizeMax));
+		assertThat(directSize, is(TM_RESOURCE_SPEC.getTaskOffHeapSize().add(TM_RESOURCE_SPEC.getShuffleMemSize())));
```
off heap managed memory is missing

xintongsong replied:
That uses unsafe memory, which should not be accounted for either `-Xmx` nor `-XX:MaxDirectMemorySize`.

tillrohrmann replied:
Yes you are right. Thanks for the clarification.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -0,0 +1,323 @@
+		conf.setString(TaskManagerOptions.TOTAL_FLINK_MEMORY, totalFlinkMemorySize.getMebiBytes() + "m");
+
+		TaskExecutorResourceSpec taskExecutorResourceSpec = TaskExecutorResourceUtils.resourceSpecFromConfig(conf);
+		assertThat(taskExecutorResourceSpec.getTotalFlinkMemorySize(), is(totalFlinkMemorySize));
```
Shouldn't we also assert that we have configured the remaining values correctly?

xintongsong replied:
`TaskExecutorResourceSpec#getTotalFlinkMemorySize` gets the Total Flink Memory size by adding up individual memory sizes. So this test case only need to validate that the added up Total Flink Memory size equals to the configured value.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -0,0 +1,323 @@
+		conf.setString(TaskManagerOptions.TOTAL_PROCESS_MEMORY, totalProcessMemorySize.getMebiBytes() + "m");
+
+		TaskExecutorResourceSpec taskExecutorResourceSpec = TaskExecutorResourceUtils.resourceSpecFromConfig(conf);
+		assertThat(taskExecutorResourceSpec.getTotalProcessMemorySize(), is(totalProcessMemorySize));
```
Same here

xintongsong replied:
Same for `TaskExecutorResourceSpec#getProcessFlinkMemorySize`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -0,0 +1,323 @@
+/**
+ * Tests for {@link TaskExecutorResourceUtils}.
+ */
+public class TaskExecutorResourceUtilsTest extends TestLogger {
```
There are no tests verifying that a misconfigured cluster fails with the appropriate exception.

tillrohrmann replied:
What about backwards compatibility tests?

xintongsong replied:
I'll add tests for failure cases and backwards compatibilities.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -0,0 +1,323 @@
+
+		// validate only in configuration with explicit task heap and managed memory size,
+		// in other cases, heap memory is derived from total flink memory, thus the configured value is not used by design
+		validateInConfigWithExplicitTaskHeapAndManagedMem(conf, taskExecutorResourceSpec -> assertThat(taskExecutorResourceSpec.getTaskHeapSize(), is(taskHeapSize)));
```
I find it confusing that certain configuration values are only valid in certain scenarios.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,333 @@
+				+ "), Shuffle Memory (" + shuffleMemorySize.toString()
+				+ ") and Managed Memory (" + managedMemorySize.toString()
+				+ ") exceed configured Total Flink Memory (" + totalFlinkMemorySize.toString() + ").");
+		final MemorySize taskHeapMemorySize = totalFlinkMemorySize.subtract(totalFlinkExceptTaskHeapMemorySize);
```
Currently in this PR, `taskmanager.memory.task.heap.size` will be ignored in such cases.

I also noticed your concern during the implementation, that the user explicitly configured task heap memory size is not taking effect only because managed memory size is not explicitly configured at the same time, which is quite not intuitive to the users.

However, if we want to respect user's explicit configurations as much as possible, then it would be a question that which implicit configuration should be overwritten. E.g., if the user explicitly configured total flink memory size and task heap memory size, and the explicitly configured task heap memory size is larger than derived from total flink memory size, the which memory pool should we shrink to guarantee the explicit configured task heap memory size? Leaving alone task off heap memory and framework heap memory (which both have default absolute values), should we prefer shrinking managed memory or shuffle memory? And do we respect the implicit min sizes of managed/shuffle memory? What if shrinking either of managed/shuffle memory can not give enough memory to the explicitly configured task heap memory (wrt the min sizes), so that we need to shrink both of them? How many should we shrink on each of them?

It seems to me that the calculation logics can be complicated if we try to respect all explicit user configurations, which may also not good for user to understand and for maintainability. Thus before we decide how to deal with this tradeoff, I implemented this PR strictly following FLIP-49 design doc.

tillrohrmann replied:
I think the logic should be to try to minimize other pool budgets if the minimum configured value allows it and if this is not enough, then we fail. Which pool is shrunk, I would not give a guarantee. It should also be ok to pick an arbitrary one because otherwise the user would have configured a different min value.

xintongsong commented:
Thanks for the review, @tillrohrmann.
It seems that the two main concerns are explicit configured values and backwards compatibility.

Regarding explicit configured values, here are what I think:
- We should always respect explicit configured values
- We should also always respect the default absolute values of the following config options, if not explicitly configured
  - Framework Heap Memory
  - Task Off-Heap Memory
  - Min/Max Shuffle Memory
  - JVM Metaspace
  - Min/Max JVM Overhead
- If Task Heap Memory and Managed Memory are not both explicitly configured, and Total Flink Memory is either explicitly configured or can be derived from explicitly configured Total ProcessMemory
  - If none of Task Heap Memory and Managed Memory is configured, or only Managed Memory is configured but not Task Heap Memory, we follow the current calculation logic.
  - If only Task Heap Memory is configured, but not Managed Memory
    - If Managed Memory fraction is explicitly configured, we first derive Managed Memory size from the fraction, then leave the remaining (Total Flink Memory minus Framework Heap Memory, Task Heap/Off-Heap Memory and Managed Memory) to Shuffle Memory. If the derived Shuffle Memory size is not in the (explicitly configured or default) Min/Max Shuffle Memory, we fail.
    - If Managed Memory fraction is not configured, we first derive Shuffle Memory (from min/max/fraction), then leave the remaining to Managed Memory.

Regarding backwards compatibility, the following config options are not supported in the current PR:
- `taskmanager.memory.preallocate`: Pre-allocation is no longer supported.
- `taskmanager.network.numberOfBuffers`: I talked to @zhijiangW about this. I think we can still support this config option as a fallback. This option will only take effect when 1) it is explicitly configured and 2) none of `taskmanager.memory.shuffle.min/max/fraction` or `taskmanager.network.memory.min/max/fraction` is configured. When it takes effect, we consider Shuffle Memory Size is explicitly configured by user, and derive it from page size * buffer num.
- `taskmanager.memory.fraction` (the legacy managed memory fraction): I'm in favor of removing this config option, for the following reasons.
  - The definition of the legacy fraction and new fraction are different. Keeping both of them can be confusing. 
    - legacy fraction = managed memory / (total flink memory - shuffle memory)
    - new fraction = managed memory / total flink memory
  - The calculation can be complicated. Computing managed memory size from legacy fraction depends on the shuffle memory size, while sometimes shuffle memory size also depends on the managed memory size. Even we derive managed memory and network memory together, from total flink memory and the sum of managed memory fraction and network memory fraction, shuffle memory can still be overwritten by its min/max limit, causing the actual fraction that managed memory takes changes.

What do you think? I'd like to hear your feedbacks on above two issues before updating the PR. Thank you.

tillrohrmann commented:
Yes, all explicitly configured values should be respected.

I agree, the default values of

* Framework Heap Memory
* Task Off-Heap Memory
* Min/Max Shuffle Memory
* JVM Metaspace
* Min/Max JVM Overhead

should be respected.

In the end, we are trying to solve a linear programming problem for which there are known methods how to solve it. The problem should have the following form

```
0 = jvm_overhead + jvm_metaspace + framework + task_heap + task_off_heap + shuffle + managed_memory - process_memory
min_overhead <= jvm_overhead
max_overhead >= jvm_overhead
min_shuffle <= shuffle
max_shuffle >= shuffle
0 = fraction_overhead * (jvm_overhead + jvm_metaspace + framework + task_heap + task_off_heap + shuffle + managed_memory) - jvm_overhead
```

If `shuffle` has not been set, then we have additionally `0 = fraction_shuffle * (framework + task_heap + task_off_heap + shuffle + managed_memory) - shuffle` 

There might be some shortcuts given the particulars of this problem. But I'm not sure whether it is exactly as you've described. I would suggest to rely on known approaches.

Concerning the backwards compatibility plan, I agree that considering the legacy `fraction`, turns the linear problem into a non-linear problem. Hence, I would be ok with dropping support for it. I guess we should add a big release note and log a warning when using the legacy `fraction` option.

tillrohrmann commented:
One way to solve the problem could be

```
final int numVariables = 8;
		final double[] c = new double[numVariables];
		Arrays.fill(c, 1.0);

		LinearObjectiveFunction f = new LinearObjectiveFunction(c, 0);
		Collection<LinearConstraint> constraints = new ArrayList<>();

		// JVM_overhead, jvm_metaspace, framework, task_heap, task_off_heap, shuffle, managed_memory, total process memory
		final double[] overallMemory = {1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, -1.0};
		final double[] jvmOverhead = {1.0, 0, 0, 0, 0, 0, 0, 0};
		final double[] jvmMetaspace = {0, 1.0, 0, 0, 0, 0, 0, 0};
		final double[] framework = {0, 0, 1.0, 0, 0, 0, 0, 0};
		final double[] taskOffHeap = {0, 0, 0, 0, 1.0, 0, 0, 0};
		final double minOverhead = 10;
		final double maxOverhead = 100;
		final double[] shuffle = {0, 0, 0, 0, 0, 1.0, 0, 0};
		final double minShuffle = 10;
		final double maxShuffle = 100;
		final double overheadFraction = 0.1;
		final double[] jvmOverheadRelationship = {-1.0, overheadFraction, overheadFraction, overheadFraction, overheadFraction, overheadFraction, overheadFraction, 0};
		final double shuffleFraction = 0.2;
		final double[] shuffleRelationship = {0, 0, shuffleFraction, shuffleFraction, shuffleFraction, -1, overheadFraction, 0};
		final double jvmMetaspaceValue = 192;
		final double frameworkValue = 192;
		final double taskOffHeapValue = 0;
		constraints.add(new LinearConstraint(overallMemory, Relationship.EQ, 0));
		constraints.add(new LinearConstraint(jvmOverhead, Relationship.GEQ, minOverhead));
		constraints.add(new LinearConstraint(jvmOverhead, Relationship.LEQ, maxOverhead));
		constraints.add(new LinearConstraint(shuffle, Relationship.GEQ, minShuffle));
		constraints.add(new LinearConstraint(shuffle, Relationship.LEQ, maxShuffle));
		constraints.add(new LinearConstraint(jvmOverheadRelationship, Relationship.EQ, 0));
		constraints.add(new LinearConstraint(shuffleRelationship, Relationship.EQ, 0));
		constraints.add(new LinearConstraint(jvmMetaspace, Relationship.EQ, jvmMetaspaceValue));
		constraints.add(new LinearConstraint(framework, Relationship.EQ, frameworkValue));
		constraints.add(new LinearConstraint(taskOffHeap, Relationship.EQ, taskOffHeapValue));

		SimplexSolver solver = new SimplexSolver();
		PointValuePair optSolution = solver.optimize(
			new MaxIter(100),
			f,
			new LinearConstraintSet(constraints),
			GoalType.MAXIMIZE,
			new NonNegativeConstraint(true));
```

xintongsong commented:
@tillrohrmann,

I agree that the memory configurations can be modeled as a linear programming problem, but I'm not sure whether we should do this. I'm in favor of derive the memory sizes manually, like how it is done currently with a if-else branches.

I have some concerns on using known linear programming solvers.
- We are limited by the double type precision, and cannot use BigDecimal in case of rounding errors.
- If there is a `NoFeasibleSolutionException`, we may not be able to provide helpful information to help the user fix their configurations. E.g., which values are not adding up, and which values can potentially be increased/decreased to fix the problem.

On the other hand, manually calculating the memory sizes (one after another) may not solve the exactly same linear programming problem (it may not explore the entire solution space due to early deciding some of the memory sizes). But maybe this is not a bad thing. I think it would be easier for user to understand how a particular memory size is calculated (less dependencies on others), and how to change/ fix it in case of conflicts.

I think it would be good that given the explicitly configured values, we try to find a feasible solution by tuning the none specified values. But I think it's not necessary that as long as there's a feasible solution we can always find it, especially at the cost of the two concerns mentioned above.

What do you think?

tillrohrmann commented:
Ok, I think you are right that producing a meaningful error message can be hard based on the linear program's solvability result @xintongsong.

If I understood the configuration options correctly, then we need at least one of the following options to be configured: `TOTAL_PROCESS_MEMORY`, `TOTAL_FLINK_MEMORY` or `TASK_HEAP_MEMORY`. If none is set, then we can fail and tell the user about it.

If at least one of these options is set, one could theoretically compute all the remaining values (either based on fraction or concrete size value). However, it is not guaranteed that the result given other explicitly configured values is actually valid. I guess this actually only affects the above mentioned options. Hence, it could be enough to simply check at the end whether the explicit values of `TOTAL_PROCESS_MEMORY`, `TOTAL_FLINK_MEMORY` or `TASK_HEAP_MEMORY` match the result.

Long story short, I'm ok with using the if-else approach since it allows to give better error messages. For me, the important bit is that we use all explicitly configured values and that we do a sanity check.

xintongsong commented:
Great, thanks for the feedbacks.
I'll update the PR ASAP.

xintongsong commented:
@tillrohrmann @KarmaGYZ 
I've updated the PR with the following changes. Please take a look. Thanks~
- Address comments
  - Adjust memory calculation logics to respect explicit configured values
  - Support backwards compatibility for legacy configs
  - Add test cases for failure cases and backwards compatibility
  - Other minor comments
- Rebase to the latest master for conflict resolving.
- Reorganized commits and updated PR description.

The PR has already passed travis test. 
https://travis-ci.org/xintongsong/flink/builds/594393921

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -260,6 +263,189 @@
+	 * Total Process Memory size for the TaskExecutors.
+	 */
+	public static final ConfigOption<String> TOTAL_PROCESS_MEMORY =
+		key("taskmanager.memory.total-process.size")
```
Maybe
```suggestion
		key("taskmanager.memory.process.size")
```

xintongsong replied:
I'm in favor of keep the prefix 'total-', to indicate that this is a coarser level config option that is equivalent to the sum of other options.

tillrohrmann replied:
ok

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -260,6 +263,189 @@
+	 */
+	@Documentation.CommonOption(position = Documentation.CommonOption.POSITION_MEMORY)
+	public static final ConfigOption<String> TOTAL_FLINK_MEMORY =
+		key("taskmanager.memory.total-flink.size")
```
Maybe

```suggestion
		key("taskmanager.memory.flink.size")
```

tillrohrmann replied:
Maybe not, if `flink` alone could be confused with `framework`.

xintongsong replied:
Same here.

tillrohrmann replied:
ok

tillrohrmann commented on flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/itcases/HAQueryableStateFsBackendITCase.java:
```diff
@@ -96,7 +96,7 @@ private static Configuration getConfig() throws Exception {
 		Configuration config = new Configuration();
 		config.setBoolean(QueryableStateOptions.ENABLE_QUERYABLE_STATE_PROXY_SERVER, true);
-		config.setString(TaskManagerOptions.MANAGED_MEMORY_SIZE, "4m");
+		config.setString(TaskManagerOptions.LEGACY_MANAGED_MEMORY_SIZE, "4m");
```
Should we replace these config keys with the non deprecated version?

xintongsong replied:
Not until we remove the feature option and clean up the legacy codes.
Currently, FLIP-49 is disabled by default. If we use the new key here, it will be ignored by the legacy code path.

tillrohrmann replied:
ok, makes sense.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceSpec.java:
```diff
@@ -0,0 +1,158 @@
+ *            │   ┌─────────────────────────────┐
+ *            └─ ││        JVM Overhead         ││
+ *                └─────────────────────────────┘
+ *               └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
```
This is duplicated information. I would keep it in one place and link to it. Otherwise I fear that they might go out of sync at some point in time.

xintongsong replied:
Make sense. Then I would say we keep the information in `TaskExecutorResourceSpec` and remove it from `TaskExecutorResourceUtils`.

tillrohrmann replied:
Ok, sounds good.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,659 @@
+	// ------------------------------------------------------------------------
+
+	public static TaskExecutorResourceSpec resourceSpecFromConfig(final Configuration config) {
+		if (isTaskHeapMemorySizeExplicitlyConfigured(config) && isManagedMemorySizeExplicitlyConfigured(config)) {
```
Why does the managed memory needs to be explicitly configured? Can't we calculate it via `MANAGED_MEMORY_FRACTION` if it's not set?

xintongsong replied:
We can derive managed memory size from its fraction in either of the following two ways:
- `managed_size = total_flink_size * managed_fraction`
- `managed_size = remaining_flink_internal_size / (1 - managed_fraction) * managed_fraction`

If total flink memory is set or can be derived from total process memory, we use the first equation to derive managed memory size in later 'else-if' branches.

For this 'if' branch, if managed memory size is not explicitly configured and we want to derive it from its fraction, we will have to use the second equation. In that case, we need `remaining_flink_internal_size` to be determinate, which may not always be the case given that shuffle memory size can also not be explicitly configured.

There are some alternatives that I can think of, but I'm not sure whether it's necessary to go with either of them, given the complexity for user to understand and for the maintainability.
- We can derive managed and shuffle memory size together, with the following equations
  - `managed_and_shuffle_size = remaining_flink_internal_size / (1 - managed_fraction - shuffle_fraction) * (managed_fraction + shuffle_fraction)`
  - `managed_size = managed_and_shuffle_size / (managed_fraction + shuffle_fraction) * managed_fraction`
  - `shuffle_size = managed_and_shuffle_size / (managed_fraction + shuffle_fraction) * shuffle_fraction`
But given that shuffle memory size also need to respect to the min/max, this approach can be even more complicated.
- We can require either managed memory size or shuffle memory size to be explicitly configured, and derive the other with its fraction. This also complicates the calculation logic by introducing more 'if-else' branches.

tillrohrmann replied:
Ok, then we say as a limitation one needs to also configure the managed memory if one has the task heap memory configured. I guess this should then be properly documented.

xintongsong replied:
Sure.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,659 @@
+			} else {
+				shuffleMemorySize = deriveShuffleMemoryWithInverseFraction(config, totalFlinkExcludeShuffleMemorySize);
+			}
+		}
```
Shuffle memory is slightly different from managed memory.
- Managed memory is either configured with exact size of derived from fraction. If the size is not set, the (default or configured) fraction is always respected.
- Shuffle memory is configured with a group of min/max/fraction. While min/max are strict constrains, fraction is respected with best effort and can be override. This aligns with how it was configured previously.

That means any value within the min/max range can be a legal shuffle memory size. If total flink memory is not configured, we use shuffle fraction to choose one proper shuffle size from the legal range. If total flink memory size is configured, we choose the value that makes the total flink memory size add up.

tillrohrmann replied:
Ah I see. Thanks for the clarification. Ok, then there is no other way around it.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,659 @@
+						+ ") exceed configured Total Flink Memory (" + totalFlinkMemorySize.toString() + ").");
+			}
+			taskHeapMemorySize = totalFlinkMemorySize.subtract(totalFlinkExcludeTaskHeapMemorySize);
+		}
```
Same reason here.
If task heap memory is configured, we comply shuffle memory size to make total flink memory add up. If task heap memory is not configured, then we decide shuffle memory size with the fraction and leave the remaining to task heap.

tillrohrmann replied:
Ok, this makes sense. Thanks for the clarification.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,659 @@
+				managedMemorySize = deriveManagedMemoryAbsoluteOrWithFraction(config, totalFlinkMemorySize);
+			}
+			shuffleMemorySize = deriveShuffleMemoryWithFraction(config, totalFlinkMemorySize);
+			sanityCheckShuffleMemory(config, shuffleMemorySize, totalFlinkMemorySize);
```
True, this is not necessary.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,659 @@
+						+ "), Task Off-Heap Memory (" + taskOffHeapMemorySize.toString()
+						+ "), Managed Memory (" + managedMemorySize.toString()
+						+ ") and Shuffle Memory (" + shuffleMemorySize.toString()
+						+ ") exceed configured Total Flink Memory (" + totalFlinkMemorySize.toString() + ").");
```
Without this check, it will fail at the next line `taskHeapMemorySize = totalFlinkMemorySize.subtract(totalFlinkExcludeTaskHeapMemorySize);`, because memory size cannot be negative. In that case, we may fail to provide helpful error message.

tillrohrmann replied:
Makes sense. Then let's keep it like this.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -0,0 +1,659 @@
+		return flinkInternalMemory;
+	}
+
+	@SuppressWarnings("deprecation")
```
Got it. I didn't know that annotations can be put inside method bodies.

xintongsong commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -0,0 +1,685 @@
+	@Test
+	public void testConfigOffHeapManagedMemoryFractionFailure() {
+		// negative off-heap managed memory fraction means not configured, if off-heap managed memory size is also not configured,
+		// legacy 'taskmanager.memory.off-heap' will be used to set managed memory to either all on-heap or all off-heap
```
My bad, it should belong to the next case.

xintongsong commented:
Thanks for the review, @tillrohrmann.
I clarified a bit on the calculation logic of shuffle memory. I think the complexity is a fair price we pay for backwards compatibility on shuffle memory min/max/fraction. What do you think?
Will update the PR ASAP.

tillrohrmann commented:
> Thanks for the review, @tillrohrmann.
> I clarified a bit on the calculation logic of shuffle memory. I think the complexity is a fair price we pay for backwards compatibility on shuffle memory min/max/fraction. What do you think?

Sounds good to me. Thanks a lot for the clarification @xintongsong. Ping me once the PR is updated and rebased, then I can take a final look and merge it.

xintongsong commented:
Thanks @tillrohrmann.
PR is updated and rebased. Travis passed: https://travis-ci.org/xintongsong/flink/builds/596889510
